### ISSUE TITLE
Question about the self-defined package in the golang-middleware template

### ISSUE THREAD

**Issue by FlankMiao** [openfaas/faas#917]
I use the golang-middleware template to build a function like Test.
I found that if I defined my own package like helloworld in the Test dir, when I build the function, I got error about can,t find the helloworld package.
like this:
**function/run.go:9:2: cannot find package "github.com/docker/docker/api/types" in any of:
        /go/src/handler/function/vendor/github.com/docker/docker/api/types (vendor tree)
        /usr/local/go/src/github.com/docker/docker/api/types (from $GOROOT)
        /go/src/github.com/docker/docker/api/types (from $GOPATH)**
However the docker dependency does exist in the gopath dir, I,m confused.
In my a single main.go test, It,s ok.
So it means that I can,t define and import my own package and have to place them in the same dir as handler.go?
And why when I build, it can,t find the dependency in the gopath?

**Comment by alexellis** [comment]
If you want to make use of the Docker SDK for instance then you need to *vendor* the dependency.

Here's a full worked-example of using vendoring - https://blog.alexellis.io/serverless-golang-with-openfaas/

Alex

**Comment by alexellis** [comment]
Feel free to keep commenting.

**Comment by martindekov** [comment]
Hello @FlankMiao . Easiest way to work with this is to create vendor folder(as you did) next to your handler and put all your dependencies there.The difference is that you put directly your dependencies in the folder. You can use `go dep` tool to integrate the dependencies from your system nest the folders inside the vendor folder like `"github.com/docker/docker/api/types"` or manually put your `"/api/types"` inside the vendor folder and import as 
```
import "/api/types"
```
Building is replacing the template handler folder with the content of your handler folder. The template don't actually copy the dependencies from your $GOROOT/GOPATH but rather only the content inside the handler folder(if vendor exist go will know in that folder there are dependencies).

TL:DR:
1. Create vendor folder next to handler.go
2. Put all of your dependencies there

**Comment by FlankMiao** [comment]
@martindekov  Thanks , I got it. The import dir in this is not the same as $gopath. :)
